### PR TITLE
Add junit Pipeline step and junitWrapper Pipeline step.

### PR DESCRIPTION
**Pull request by abayer**
More testing to be done, needs documentation, needs better error propagation.

- [x] Write the actual step.
- [ ] Help documentation for the global variable.
- [ ] Figure out how to ensure that nested errors still get surfaced even when the `JUnitResultArchiver` `step` fails - right now they're swallowed.

cc @reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
See #9.

**Comment by abayer** [comment]
Meaning what? =)

**Comment by jglick** [comment]
Meaning this has been done before, I had some reservations, read up on existing approaches and all associated material.

**Comment by abayer** [comment]
So you're saying not to do it in the junit plugin, even with optional dependencies and extensions?

**Comment by jglick** [comment]
No optional dependencies. If you want to define a step, define it. There is a reason `workflow-step-api` provides no GUI: so there is no cost to depending on it.

**Comment by abayer** [comment]
fwiw, tests pass fine with Java 8, since, well, Permgen.

**Comment by abayer** [comment]
@jglick - So...don't use optional dependencies, meaning don't do it in junit-plugin due to 1.x bundling, meaning do it in a new plugin?

**Comment by abayer** [comment]
Closing this in favor of the `junit-pipeline-steps` plugin that's currently waiting to be moved to the `jenkinsci` org.

**Comment by jglick** [comment]
> don't use optional dependencies, meaning don't do it in junit-plugin due to 1.x bundling, meaning do it in a new plugin?

No, meaning simply add the `workflow-step-api` dependency to the `junit` plugin. Why not?

**Comment by abayer** [comment]
I am so confused. =) I thought that adding the `workflow-step-api` dependency would create problems in 1.x due to bundling?

**Comment by jglick** [comment]
> I thought that adding the `workflow-step-api` dependency would create problems in 1.x due to bundling?

Well, we are not going to produce any more 1.x releases, so nothing you do in a future release of this plugin is going to matter for 1.x bundling, right?

**Comment by abayer** [comment]
Valid point, that. I'll reopen this. =)

File: src/main/resources/hudson/tasks/junit/pipeline/JUnitWrapperPipelineStep.groovy [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.jenkinsci.plugins.workflow.cps.CpsScript
+
+class JUnitWrapperPipelineStep implements Serializable {
+    CpsScript script
+
+    JUnitWrapperPipelineStep(CpsScript script) {
+        this.script = script
+    }
+
+    def call(Map args, Closure closure) {
```
Comment (jglick) [inline src/main/resources/hudson/tasks/junit/pipeline/JUnitWrapperPipelineStep.groovy]: Uh, what?? Just make a plain old `Step`.

In fact this DSL is just sugar over the existing `SimpleBuildStep` and so is less valuable than the original PR, which

* allowed you to inspect results in Groovy code and make decisions based on them
* allowed you to decide whether or not to set the build status

File: src/main/resources/hudson/tasks/junit/pipeline/JUnitPipelineStep.groovy [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        this.script = script
+    }
+
+    def call(Map args) {
+        String testResults = args.containsKey("testResults") ? args.testResults : ""
+        boolean keepLongStdio = args.containsKey("keepLongStdio") ? args.keepLongStdio : false
+        boolean allowEmptyResults = args.containsKey("allowEmptyResults") ? args.allowEmptyResults : false
+        Double healthScaleFactor = args.containsKey("healthScaleFactor") ? args.healthScaleFactor : 1.0
+
+        script.step($class: "JUnitResultArchiver",
```
Comment (jglick) [inline src/main/resources/hudson/tasks/junit/pipeline/JUnitPipelineStep.groovy]: :-1:, this adds no value at all, other than creating a one-off DSL with no `Snippetizer` support that will be obsoleted by JENKINS-29922.

**Comment by jglick** [comment]
Sorry but :-1: on the concept here.

**Comment by abayer** [comment]
Ok, I'll revisit this later, then.

**Comment by abayer** [comment]
Closing - the new metastep syntax suffices.

**Pull request closed without merging**